### DISCUSSION TITLE
IRunnerReporter doesn't work in 2.8.1

### DISCUSSION THREAD

**Discussion by jchesshirAspire** [xunit/xunit#2968] [Question] [answered]
We've been running our tests with XUnit 2.4.1 for a while.  We are now trying to skip several versions and get up to the latest 2.8.1.  I've changed xunit and xunit.runner.utility from 2.4.1 to 2.8.1, and xunit.runner.visualstudio from 2.4.3 to 2.8.1.

Included in our personal test framework are classes that implement IRunnerReporter.  According to examples I found, it registers an XUnitMessageSink object.  XUnitMessageSink has been inheriting from IMessageSink and IMessageSinkWithTypes.  After the upgrade, I found that the compiler also required it to inherit from LongLivedMarshalByRefObject for some reason.

But when I run my tests now, I'm finding that is not finding and instantiating the IRunnerReporter.  This in turn, means that my XUnitMessageSink class is not able to detect events coming from XUnit.  What more do I need to do to get this working again?

**Comment by bradwilson** [comment]
> After the upgrade, I found that the compiler also required it to inherit from LongLivedMarshalByRefObject for some reason.

This has always been true. However, it only recently got surfaced to you because we added analyzers to highlight this fact.

> When I run my tests now, I'm finding that is not finding and instantiating the IRunnerReporter.

Can you provide a repro project with the reporter?

**Reply by jchesshirAspire** [reply, marked as answer]
> > After the upgrade, I found that the compiler also required it to inherit from LongLivedMarshalByRefObject for some reason.
> 
> This has always been true. However, it only recently got surfaced to you because we added analyzers to highlight this fact.
> 

Well, it always worked for me before it became an error! 😄 

> > When I run my tests now, I'm finding that is not finding and instantiating the IRunnerReporter.
> 
> Can you provide a repro project with the reporter?

No need.  I figured it out.  Turns out this was predicted in the [2.5.0 Release Notes](https://xunit.net/releases/v2/2.5.0) in the second bullet point in the Extensibility section.  All I had to do to fix it was to to pull the class implementing IRunnerReporter into a new project that has "reporters" in its name, make it depend on my base test project to register the message sink and listener, and have the test projects that need the listener depend on the "reporters" as well as the base test project.  Then it worked just like before.

**Reply by bradwilson** [reply]
> Well, it always worked for me before it became an error! 😄

If you don't use .NET Framework, then the issue likely never comes up. It doesn't change the "requirement", but the "fake" implementation of `LongLivedMarshalByRefObject` when you're not running on .NET Framework shows it to just be a no-op: https://github.com/xunit/xunit/blob/80a22946275918c653a43ebaee6a82c2541b4195/src/common/LongLivedMarshalByRefObject.cs#L57-L68

**Comment by DevOnRun** [comment]
@jchesshirAspire, Can you please share a sample project with suggested workaround i.e. `implementing IRunnerReporter into a new project that has "reporters" in its name`
Also, @bradwilson, Is XUnit working for permanent fix for this to avoid above mentioned workaround?

**Reply by bradwilson** [reply]
No, there are no changes planned for this.
